src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -454,7 +476,13 @@ void MergeTreeData::checkProperties(
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Primary key must be a prefix of the sorting key, but its length: "
             "{} is greater than the sorting key length: {}", primary_key_size, sorting_key_size);
 
-    NameSet primary_key_columns_set;
+    bool allow_suspicious_indices = getSettings()->allow_suspicious_indices;
+    if (local_context)
+        allow_suspicious_indices = local_context->getSettingsRef().allow_suspicious_indices;
```

Comment:
Why can it be a nullptr? Should we get the setting from the global context in this case?

Reply:
> Global context = merge tree settings
> Local context = global settings.

But why? Do we have other settings with such behavior?

The problem with `if (local_context)` is that changing the setting (even for the default profile) will have no effect on background operations. That's confusing and inconsistent with other settings (like `allow_nondeterministic_mutations` or `allow_suspicious_codecs`)

Reply:
> will have no effect on background operations.

Another question is why we recheck indices in background operations like applying `ALTER_METADATA`. It should have been checked on the initiator before committing `ALTER_METADATA`. It does not make sense to recheck it when applying committed changes